libtsuba/include/tsuba/Cache.h

```diff
@@ -0,0 +1,182 @@
+#ifndef KATANA_LIBTSUBA_TSUBA_CACHE_H_
+#define KATANA_LIBTSUBA_TSUBA_CACHE_H_
+
+// This is single threaded only, it is not intended to store large objects,
+// but rather metadata (e.g., a shared_ptr to a property column).
+
+// The problem witchel had implementing a multi-threaded version using
+// parallel-hashmap is a lock ordering problem.  parallel-hashmap 1.33 allows
+// execution of code with the write lock held, only when modifying, not when adding an
+// element.  We have to modify the LRU list when eviciting an element, so the natural
+// lock ordering is parallel-hashmap write lock, then list lock.  But without a way to
+// execute insert code with the parallel-hashmap write lock held, it seemed like there
+// would be some form of race condition.
+
+#include <list>
+#include <optional>
+#include <string>
+
+#include <boost/container_hash/hash.hpp>
+
+#include "katana/ArrowInterchange.h"
+#include "katana/ConcurrentMap.h"
+#include "katana/Logging.h"
+
+namespace tsuba {
+
+struct KATANA_EXPORT KeyBase {
+  enum class NodeEdge { kNode, kEdge };
+  bool operator==(const KeyBase& o) const {
+    return node_edge_ == o.node_edge_ && name_ == o.name_;
+  }
+
+  NodeEdge node_edge_;
+  std::string name_;
+
+  struct Hash {
+    std::size_t operator()(const KeyBase& key) const {
+      using boost::hash_combine;
+      using boost::hash_value;
+
+      std::size_t seed = 0;
+      hash_combine(seed, hash_value(key.node_edge_));
+      hash_combine(seed, hash_value(key.name_));
+
+      // Return the result.
+      return seed;
+    }
+  };
+};
+
+template <typename Value>
+class KATANA_EXPORT Cache {
```

Comment:
I thought about doing that, but then thought why not leave it generic.  But sure, I can make it `LRUCache`.

Reply:
Hmm, I started doing this and I don't like it.  The replacement policy feels like an implementation detail that should stay out of the name.

Reply:
That may be true, but the implementation of different policies will lead to varying fields in the implementation class and the current map + linked list implementation may not be suitable for let's say prioritized replacement.

If you want a general interface you try a cache base class, but as you note, the current interface is only appropriate for things that are conveniently copyable. And with a base class, you'd have a hard time expressing the template Value and possible Key parameter.

With templates, you can only achieve textual/static polymorphism (i.e., if it happens to have the same name functions and the return values are interchangable, you can replace one cache class for another) versus the dynamic polymorphism of virtual. The way you indicate substitutability with templates (c.f. virtual base class) is simply to have multiple classes with the same function name and signature (C++ concepts just formalizes this). 

So, with templates the way you indicate there are bunch of interchangable cache classes is simply by convention.

```cpp

template<typename T>
class LRUCache {
public:
  T Get();
};

template<typename T>
class FancyCache {
public:
  T Get();
};

template <typename T>
using MyCache = FancyCache;

MyCache cache;
```

We can swap out cache implementations because they all have the same function names.

vs

```cpp
class Cache {
public:
  virtual Object Get() = 0;
};

class LRUCache: public Cache {
public:
  Object Get() override {};
};

Cache cache = new LRUCache();
```

it is not advisable to mix these two paradigms.

Reply:
Interesting and appreciated.   I have tried to implement the template form of polymorphism for the Cache Key.

Some replacement policies might not work with the current structure, but some, like MRU or segmented LRU will.